lib/prelude.dx

```diff
@@ -983,6 +990,32 @@ instance Monoid (List a)
         True  -> xs.(unsafeFromOrdinal _ i')
         False -> ys.(unsafeFromOrdinal _ (i' - nx))
 
+def ListMonoid (a:Type) : Monoid (List a) =
+  A = a -- XXX: Typing `Monoid a` below would quantify it over a,
+        --      which we don't want.
+  named-instance result : Monoid (List A)
+    mempty = mempty
+    mcombine = mcombine
+  result
+
+def append [AccumMonoid h (List a)]
+  (list: Ref h (List a)) (x:a) : {Accum h} Unit =
+    list += AsList 1 [x]
+
+def filter (condition:a->Bool) (xs:n=>a) : List a =
+  yieldAccum (ListMonoid a) \list.
+    for i.
+      if condition xs.i
+        then append list xs.i
```

Comment:
I kind of wish we could write this in terms of `argFilter` composed with indexing and have Dex emit exactly this code, but I doubt our optimization passes are strong enough 😕